rapid_response_xblock/static/js/src/rapid.js

```diff
@@ -390,35 +378,120 @@
+
+      $buttonsRow.toggleClass('hidden', false);
+      $problemButton.text((state.is_open ? "Close" : "Open") + " Problem Now");
+
+      var totalSeconds = getSecondsSinceRunOpened();
+      var pollSeconds = 0, pollMinutes = 0;
+      if (state.is_open) {
+        pollSeconds = totalSeconds % 60;
+        pollMinutes = Math.floor(totalSeconds / 60);
+      }
+      $timer.text(pollMinutes + "m : " + pollSeconds + "s");
+      $timerSpinner.toggleClass('hidden', !state.is_open);
+      $timerSpinnerText.toggleClass('hidden', !state.is_open);
+    }
+
+    /**
+     * Update the timer every second
+     */
+    function updateTimer() {
+      if (state.is_open) {
+        setTimeout(updateTimer, TIMER_MILLIS);
+      }
+      render();
+    }
+
     /**
      * Read from the responses API and put the new value in the rendering state.
      * If the problem is open, schedule another poll using this function.
      */
     function pollForResponses() {
-      $.get(responsesUrl).then(function(newState) {
-        state = _.assign({}, state, newState);
+      if (state.is_open) {
+        setTimeout(pollForResponses, POLLING_MILLIS);
+      }
+
+      // make sure this updates at least once a second
+      if (state.isPolling) {
+        // API call is still in progress
+        return;
+      }
 
+      state.isPolling = true;
+      $.get(responsesUrl).then(function(newState) {
+        state = _.assign({}, state, newState, {
+          isPolling: false,
+          lastPoll: moment()
+        });
         render();
-        if (state.is_open) {
-          setTimeout(pollForResponses, POLLING_MILLIS);
-        }
-      }).fail(function () {
-        // TODO: try again?
-        console.error("Error retrieving response data");
       });
     }
 
     $(function($) { // onLoad
       var block = $element.find(rapidTopLevelSel);
-      _.assign(state, {
-        is_open: block.attr('data-open') === 'True',
-        is_staff: block.attr('data-staff') === 'True'
+      state.is_open = block.attr('data-open') === 'True';
+
+      var $rapidBlockContent = $element.find(rapidBlockContentSel);
+      $rapidBlockContent.find(problemStatusBtnSel).click(function() {
+        // disable the button temporarily to prevent double clicks
+        $rapidBlockContent.find(problemStatusBtnSel).prop("disabled", true);
+        $.post(toggleStatusUrl).then(
+          function(newState) {
+            $rapidBlockContent.find(problemStatusBtnSel).prop("disabled", false);
+
+            // Selected runs should be reset when the open status is changed
+            state = _.assign({}, state, newState, {
+              selectedRuns: [null]
+            });
+
+            if (state.is_open) {
+              pollForResponses();
+              updateTimer();
+            }
+
+            render();
+          }
+        );
       });
+
       render();
 
-      if (state.is_staff) {
-        pollForResponses();
-      }
+      pollForResponses();
+      updateTimer();
```

Comment:
This logic seems a little off to me. The way I'm understanding this, the logic at page load should be roughly this (by 'controls' I mean the button and timer):

- If `state.open` == true, the controls and the response polling should be started
- If `state.open` == false, responses should be fetched and the most recent response should be rendered
- Regardless of `state.open`, the controls should be rendered, and we want to fetch responses and render those as soon as they're returned

I think `updateTimer` and `pollForResponses` are somewhat misleading and that's what I'm getting hung up on. I think they're doing too much, and should both just set a function to be called after a certain interval. I also think the state variable names could be a bit more helpful. Something like this:

```javascript
function updateTimer() {
  setTimeout(renderControls, TIMER_MILLIS);
}

function pollForResponses() {
  setTimeout(fetchResponsesAndRender, POLLING_MILLIS);
}

function fetchResponsesAndRender() {
  if (state.isFetchingResponses) {
    // API call is still in progress
    return;
  }
  state.isFetchingResponses = true;
  $.get(responsesUrl).then(function(newState) {
    state = _.assign({}, state, newState, {
      isFetchingResponses: false,
      lastFetch: moment()
    });
    render();
  });
}

$(function($) { // onLoad
  //...
  $rapidBlockContent.find(problemStatusBtnSel).click(function() { 
    //...
    renderControls();
    if (state.is_open) {
      pollForResponses();
      updateTimer();
    }
  });

  renderControls();
  fetchResponsesAndRender();
  if (state.is_open) {
    pollForResponses();
    updateTimer();
  }
});
```

Reply:
It looks like `pollForResponses` and `updateTimer` would update once via `setTimeout` but I'm not sure where they would get called repeatedly